src/charm.py

```diff
@@ -0,0 +1,345 @@
+        self.topology = JujuTopology.from_charm(self)
+        self.loki_cluster = LokiClusterRequirer(self)
+        self.log_forwarder = ManualLogForwarder(
+            charm=self,
+            loki_endpoints=self.loki_cluster.get_loki_endpoints(),
+            refresh_events=[
+                self.on["loki-cluster"].relation_joined,
+                self.on["loki-cluster"].relation_changed,
+            ],
+        )
+        self.unit.set_ports(self._port)
+
+        # === EVENT HANDLER REGISTRATION === #
+        self.framework.observe(self.on.loki_pebble_ready, self._on_pebble_ready)  # pyright: ignore
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        self.framework.observe(self.loki_cluster.on.config_received, self._on_loki_config_received)
+        self.framework.observe(self.loki_cluster.on.created, self._on_loki_cluster_created)
+        self.framework.observe(
+            self.on.loki_cluster_relation_departed, self.log_forwarder.disable_logging
+        )
+        self.framework.observe(
+            self.on.loki_cluster_relation_broken, self.log_forwarder.disable_logging
+        )
+
+    # === EVENT HANDLERS === #
+    def _on_pebble_ready(self, _):
+        self.unit.set_workload_version(self._loki_version or "")
+        self._update_config()
+
+    def _on_config_changed(self, _):
+        # if the user has changed the roles, we might need to let the coordinator know
+        self._update_loki_cluster()
+
+        # if we have a config, we can start loki
+        if self.loki_cluster.get_loki_config():
+            # determine if a workload restart is necessary
+            self._update_config()
+
+    def _on_upgrade_charm(self, _):
+        self._update_loki_cluster()
+
+    def _on_collect_status(self, e: CollectStatusEvent):
+        if not self._container.can_connect():
+            e.add_status(WaitingStatus("Waiting for `loki` container"))
+        if not self.model.get_relation("loki-cluster"):
+            e.add_status(
+                BlockedStatus("Missing loki-cluster relation to a loki-coordinator charm")
+            )
+        elif not self.loki_cluster.relation:
+            e.add_status(WaitingStatus("Loki-Cluster relation not ready"))
+        if not self.loki_cluster.get_loki_config():
+            e.add_status(WaitingStatus("Waiting for coordinator to publish a loki config"))
+        if not self._loki_roles:
+            e.add_status(BlockedStatus("No roles assigned: please configure some roles"))
+        e.add_status(ActiveStatus(""))
+
+    def _on_loki_cluster_created(self, _):
+        self._update_loki_cluster()
+
+    def _on_loki_config_received(self, _e: ConfigReceivedEvent):
+        self._update_config()
+
+    # === PROPERTIES === #
+
+    @property
+    def _pebble_layer(self):
```

Comment:
strongly advise to move all loki-workload stuff onto a separate `loki.py::Loki` object. 
it works really well
see:
https://github.com/canonical/tempo-worker-k8s-operator/tree/main/src
or
https://github.com/jnsgruk/zinc-k8s-operator/tree/main/src